src/order.c

```diff
@@ -479,6 +525,60 @@ SEXP vec_order_info_impl(SEXP x,
     r_list_poke(out, 2, r_int((int) p_group_info->max_group_size));
   }
 
+  if (appearance) {
+    struct group_info* p_group_info = groups_current(p_group_infos);
+
+    const r_ssize n_groups = p_group_info->n_groups;
+    const int* p_sizes = p_group_info->p_data;
+
+    // Order of the unique keys
+    SEXP keys = PROTECT_N(r_alloc_integer(n_groups), &n_prot);
+    int* p_keys = r_int_begin(keys);
+
+    r_ssize start = 0;
+
+    for (r_ssize i = 0; i < n_groups; ++i) {
+      p_keys[i] = p_order->p_data[start];
+      start += p_sizes[i];
+    }
+
+    // Appearance order of the unique keys
+    struct order* p_order_appearance = new_order(n_groups);
```

Comment:
The idea is to find the sorted group keys of `x` and extract their order values into `key`. And then we take that integer vector of group key order values and sort it again, which gives us a way to map back to order by appearance.

One nice thing is that we are reusing all of the memory that was already allocated for the first ordering. Because this is ordering a dense integer vector and is reusing memory, it has extremely low overhead.